pkg/parse/quote.go

```diff
@@ -20,19 +21,19 @@ func QuoteVariableName(s string) string {
 		return "''"
 	}
 
-	// Keep track of whether it is a valid (unquoted) variable name.
-	bare := true
-	for _, r := range s {
-		if !unicode.IsPrint(r) {
+	bare := true // keep track of whether it is a valid (unquoted) variable name
+	for s2 := s; len(s2) > 0; {
+		r, w := utf8.DecodeRuneInString(s2)
+		if r == utf8.RuneError || !unicode.IsPrint(r) {
```

Comment:
Only this line seems to be the necessary change in this function; iterating a string with invalid UTF-8 sequences produces utf8.RuneError too.

Please also refrain from making unnecessary stylistic changes; I prefer the existing comment format on line 23 and the empty line on line 35.

Reply:
> iterating a string with invalid UTF-8 sequences produces utf8.RuneError too 

Yes, although it technically returns `unicode.ReplacementChar` which happens (albeit presumably by design) to have the same value as `utf8.RuneError`. More importantly this formulation is, AFAICT, the preferred idiom for handling strings (and byte slices) that contain invalid UTF-8 sequences. I've read a lot of blog posts and Go issues that talk about the current behavior and the consensus opinion seems to be to use the `unicode/utf8` package in these situations rather than ranging over the string. Since we have to use the `utf8.DecodeRuneInString` pattern where we actually need to capture the invalid UTF-8 byte it's preferable to use the same pattern in places like this where we don't to ease the cognitive load on readers of the code.

> Please also refrain from making unnecessary stylistic changes; I prefer the existing comment format on line 23 and the empty line on line 35.

Understood and I'll revert that part of the change. But note that I disagree, at least in the case of the comment on line 23, because it only applies to the `bare` var; not the entire block of code and thus should be, IMHO, attached to the var declaration rather than a block comment.

Reply:
> I've read a lot of blog posts and Go issues that talk about the current behavior and the consensus opinion seems to be to use the unicode/utf8 package in these situations rather than ranging over the string.

Can you provide some reference links for that, in particular the opinion from Ian Taylor (which you mentioned in another thread)?

I haven't been able to find any GitHub issues that discuss the behavior of the `range` loop over strings that contain invalid UTF-8 sequences in general, but GitHub's search functionality is pretty bad.

> Since we have to use the utf8.DecodeRuneInString pattern where we actually need to capture the invalid UTF-8 byte it's preferable to use the same pattern in places like this where we don't to ease the cognitive load on readers of the code.

You can always index the string to get the actual byte.

Reply:
Sadly, I can't find, using a google search, the message I know I read recently (but not necessarily published recently) on this topic wherein Ian Taylor advocated for using `utf8.DecodeRuneInString` rather than ranging over the string. However, see https://github.com/golang/go/issues/28599 which proposed augmenting ranging over a string to make it easier to handle invalid UTF-8.  See also the `Map` function in *src/strings/strings.go* in the Go source for how awkward it is to deal with invalid UTF-8 using `range string`. See also all the numerous uses of `utf8.DecodeRuneInString` in the Go source when iterating over, or otherwise handling, strings.

The bottom line is that I believe a pattern using `utf8.DecodeRuneInString` is easier to understand than one based on `range string`. Which approach is more efficient is probably the more salient point. I'll admit I did not do any performance tests.

Reply:
OK, reading that issue made me understand why it's desirable to use `utf8.DecodeRuneInString` to correctly handle strings with invalid UTF-8 sequences. It is to distinguish invalid sequences from actual occurrences of U+FFFD. When an actual U+FFFD is encountered, the second return value will be 3 instead of 1.

However, the code you wrote did not check this condition, meaning that it will still fail to distinguish an invalid sequence from an actual U+FFFD, and there really is no benefit over a `range` loop if you aren't checking that.

I'll fix that and add a test case.

There may be some merit in avoiding `range` loop over strings altogether and stick to `utf8.DecodeRuneInString` when iterating strings just for consistency's sake, but I find it too cumbersome to do so.

I'll revert the places where it is not necessary to use `utf8.DecodeRuneString` to the previous `range` loop code.